File: Semestrul 4/SDI/lab07/java/TriathlonNetworking/src/main/java/triathlon/network/jsonprotocol/TriathlonServicesJsonProxy.java
package triathlon.network.jsonprotocol;

import triathlon.model.Referee;
import triathlon.model.RefereeDiscipline;
import triathlon.model.Result;
import triathlon.services.ITriathlonObserver;
import triathlon.services.ITriathlonServices;
import triathlon.services.TriathlonException;
import com.google.gson.Gson;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.logging.Logger;

public class TriathlonServicesJsonProxy implements ITriathlonServices {
    private String host;
    private int port;

    private ITriathlonObserver client;

    private Socket connection;
    private PrintWriter output;
    private BufferedReader input;
    private Gson gsonFormatter;
    private BlockingQueue<Response> responses;
    private volatile boolean finished;
    private ReaderThread readerThread;

    private final static Logger logger = Logger.getLogger(TriathlonServicesJsonProxy.class.getName());

    public TriathlonServicesJsonProxy(String host, int port) {
        this.host = host;
        this.port = port;
        responses = new LinkedBlockingQueue<>();
        gsonFormatter = new Gson();
    }

    public void closeConnection() {
        finished = true;
        try {
            if (input != null) input.close();
            if (output != null) output.close();
            if (connection != null) connection.close();
            if (readerThread != null) readerThread.interrupt();
            client = null;
        } catch (IOException e) {
            logger.warning("Error closing connection: " + e.getMessage());
        }
    }

    private void initializeConnection() throws TriathlonException {
        try {
            connection = new Socket(host, port);
            output = new PrintWriter(connection.getOutputStream(), true);
            input = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            finished = false;

            
            readerThread = new ReaderThread();
            readerThread.setDaemon(true); 
            readerThread.start();

        } catch (IOException e) {
            logger.severe("Error connecting to server: " + e.getMessage());
            throw new TriathlonException("Error connecting to server: " + e.getMessage());
        }
    }

    private void sendRequest(Request request) throws TriathlonException {
        if (connection == null || connection.isClosed()) {
            initializeConnection();
        }

        
        String requestJson = gsonFormatter.toJson(request);
        logger.info("Sending request: " + requestJson);

        synchronized (output) {
            output.println(requestJson);
            output.flush();
        }

    }

    private Response readResponse() throws TriathlonException {
        try {
            
            Response response = responses.take();
            logger.info("Received response: " + gsonFormatter.toJson(response));

            if (response.getType() == ResponseType.ERROR) {
                logger.warning("Server error: " + response.getErrorMessage());
                throw new TriathlonException("Server error: " + response.getErrorMessage());
            }

            return response;
        } catch (InterruptedException e) {
            logger.warning("Interrupted while waiting for response: " + e.getMessage());
            Thread.currentThread().interrupt();
            throw new TriathlonException("Interrupted while waiting for response");
        }
    }

    @Override
    public RefereeDiscipline login(Referee referee, ITriathlonObserver client) throws TriathlonException {
        initializeConnection();
        this.client = client;
        Request request = JsonProtocolUtils.createLoginRequest(referee);
        sendRequest(request);
        Response response = readResponse();
        return response.getRefereeDiscipline();
    }

    @Override
    public void logout(Referee referee, ITriathlonObserver client) throws TriathlonException {
        Request request = JsonProtocolUtils.createLogoutRequest(referee);
        sendRequest(request);
        Response response = readResponse();
        closeConnection();
    }

    @Override
    public List<Result> getResultsByEvent(Long eventId) throws TriathlonException {
        Request request = JsonProtocolUtils.createGetResultsByEventRequest(eventId);
        sendRequest(request);
        Response response = readResponse();
        return response.getResults();
    }

    @Override
    public void updateResult(Result result) throws TriathlonException {
        Request request = JsonProtocolUtils.createUpdateResultRequest(result);
        sendRequest(request);
        Response response = readResponse();
    }

    @Override
    public int getTotalPointsForParticipant(Long participantId, Long eventId) throws TriathlonException {
        Request request = JsonProtocolUtils.createGetTotalPointsForParticipantRequest(participantId, eventId);
        sendRequest(request);
        Response response = readResponse();
        return response.getTotalPoints();
    }

    
    private class ReaderThread extends Thread {
        @Override
        public void run() {
            logger.info("Reader thread started");
            try {
                while (!finished && !Thread.currentThread().isInterrupted()) {
                    
                    if (connection != null && !connection.isClosed()) {
                        try {
                            
                            if (input.ready()) {
                                String responseJson = input.readLine();
                                logger.fine("Reader thread read: " + responseJson);

                                
                                if (responseJson == null) {
                                    logger.warning("Received null response, server likely disconnected");
                                    throw new IOException("Server disconnected");
                                }

                                
                                Response response = gsonFormatter.fromJson(responseJson, Response.class);

                                
                                if (response.getType() == ResponseType.RESULT_UPDATED) {
                                    handleResultUpdated(response);
                                } else {
                                    
                                    responses.put(response);
                                }
                            } else {
                                
                                Thread.sleep(10);
                            }
                        } catch (InterruptedException e) {
                            logger.warning("Reader thread interrupted");
                            Thread.currentThread().interrupt();
                            break;
                        } catch (IOException e) {
                            logger.severe("IO error in reader thread: " + e.getMessage());
                            
                            if (!finished) {
                                finished = true;
                                break;
                            }
                        }
                    } else {
                        
                        Thread.sleep(100);
                    }
                }
            } catch (Exception e) {
                logger.severe("Unexpected error in reader thread: " + e.getMessage());
                e.printStackTrace();
            } finally {
                logger.info("Reader thread exiting");
            }
        }

        private void handleResultUpdated(Response response) {
            
            if (client != null && response.getResults() != null) {
                try {
                    logger.info("Notifying client about result update: " +
                            response.getResults().getFirst().getParticipant().getName() + " - " +
                            response.getResults().getFirst().getPoints() + " points");

                    client.resultsUpdated(response.getResults().getFirst());
                } catch (TriathlonException e) {
                    logger.warning("Error notifying client: " + e.getMessage());
                }
            } else {
                logger.warning("Cannot handle result update: client=" + client +
                        ", result=" + (response.getResults().getFirst() != null ? "present" : "null"));
            }
        }
    }
}

File: Semestrul 4/MPP/lab08/TriathlonNetworking/jsonprotocol/TriathlonServicesJsonProxy.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using System.Collections.Concurrent;
using Newtonsoft.Json;
using log4net;
using Triathlon.Model;
using Triathlon.Services;

namespace Triathlon.Network.JsonProtocol;

public class TriathlonServicesJsonProxy : ITriathlonServices, IDisposable
{
    private readonly string _host;
    private readonly int _port;

    private ITriathlonObserver _client;

    private TcpClient _tcpClient;
    private NetworkStream _stream;
    private StreamWriter _writer;
    private StreamReader _reader;
    private readonly BlockingCollection<Response> _responses;
    private volatile bool _finished;
    private Thread _readerThread;

    private static readonly ILog _logger = LogManager.GetLogger(typeof(TriathlonServicesJsonProxy));

    public TriathlonServicesJsonProxy(string host, int port)
    {
        _host = host;
        _port = port;
        _responses = new BlockingCollection<Response>();
    }

    public void CloseConnection()
    {
        _finished = true;
        try
        {
            if (_reader != null) _reader.Close();
            if (_writer != null) _writer.Close();
            if (_stream != null) _stream.Close();
            if (_tcpClient != null) _tcpClient.Close();

            if (_readerThread != null && _readerThread.IsAlive)
            {
                _readerThread.Interrupt();
            }

            _client = null;
        }
        catch (IOException ex)
        {
            _logger.Warn($"Error closing connection: {ex.Message}");
        }
    }

    private void InitializeConnection()
    {
        try
        {
            _tcpClient = new TcpClient();
            _tcpClient.Connect(_host, _port);

            _stream = _tcpClient.GetStream();
            _writer = new StreamWriter(_stream) { AutoFlush = true };
            _reader = new StreamReader(_stream);
            _finished = false;

            _readerThread = new Thread(ReaderThreadMethod);
            _readerThread.IsBackground = true;
            _readerThread.Start();
        }
        catch (IOException ex)
        {
            _logger.Error($"Error connecting to server: {ex.Message}");
            throw new TriathlonException($"Error connecting to server: {ex.Message}");
        }
    }

    private void SendRequest(Request request)
    {
        if (_tcpClient == null || !_tcpClient.Connected)
        {
            InitializeConnection();
        }

                 string requestJson = JsonConvert.SerializeObject(request);
        _logger.Info($"Sending request: {requestJson}");

        lock (_writer)
        {
            _writer.WriteLine(requestJson);
            _writer.Flush();
        }
    }

    private Response ReadResponse()
    {
        try
        {
                         Response response = _responses.Take();
            _logger.Info($"Received response: {JsonConvert.SerializeObject(response)}");

            if (response.Type == ResponseType.ERROR)
            {
                _logger.Warn($"Server error: {response.ErrorMessage}");
                throw new TriathlonException($"Server error: {response.ErrorMessage}");
            }

            return response;
        }
        catch (InvalidOperationException ex)
        {
            _logger.Error("Collection was closed while waiting for response", ex);
            throw new TriathlonException("Connection closed while waiting for response");
        }
        catch (ThreadInterruptedException ex)
        {
            _logger.Warn($"Interrupted while waiting for response: {ex.Message}");
            Thread.CurrentThread.Interrupt();
            throw new TriathlonException("Interrupted while waiting for response");
        }
    }

    public RefereeDiscipline Login(Referee referee, ITriathlonObserver client)
    {
        InitializeConnection();
        _client = client;
        Request request = JsonProtocolUtils.CreateLoginRequest(referee);
        SendRequest(request);
        Response response = ReadResponse();
        return response.RefereeDiscipline;
    }

    public void Logout(Referee referee, ITriathlonObserver client)
    {
        Request request = JsonProtocolUtils.CreateLogoutRequest(referee);
        SendRequest(request);
        Response response = ReadResponse();
        CloseConnection();
    }

    public List<Result> GetResultsByEvent(long eventId)
    {
        Request request = JsonProtocolUtils.CreateGetResultsByEventRequest(eventId);
        SendRequest(request);
        Response response = ReadResponse();
        return response.Results;
    }

    public void UpdateResult(Result result)
    {
        Request request = JsonProtocolUtils.CreateUpdateResultRequest(result);
        SendRequest(request);
        Response response = ReadResponse();
    }

    public int GetTotalPointsForParticipant(long participantId, long eventId)
    {
        Request request = JsonProtocolUtils.CreateGetTotalPointsForParticipantRequest(participantId, eventId);
        SendRequest(request);
        Response response = ReadResponse();
        return response.TotalPoints;
    }

         private void ReaderThreadMethod()
    {
        _logger.Info("Reader thread started");
        try
        {
            while (!_finished && Thread.CurrentThread.IsAlive)
            {
                                 if (_tcpClient != null && _tcpClient.Connected)
                {
                    try
                    {
                                                 if (_tcpClient.Available > 0 || _reader.Peek() > -1)
                        {
                            string responseJson = _reader.ReadLine();
                            _logger.Debug($"Reader thread read: {responseJson}");

                                                         if (responseJson == null)
                            {
                                _logger.Warn("Received null response, server likely disconnected");
                                throw new IOException("Server disconnected");
                            }

                                                         Response response = JsonConvert.DeserializeObject<Response>(responseJson);

                                                         if (response.Type == ResponseType.RESULT_UPDATED)
                            {
                                HandleResultUpdated(response);
                            }
                            else
                            {
                                                                 _responses.Add(response);
                            }
                        }
                        else
                        {
                                                         Thread.Sleep(10);
                        }
                    }
                    catch (ThreadInterruptedException)
                    {
                        _logger.Warn("Reader thread interrupted");
                        Thread.CurrentThread.Interrupt();
                        break;
                    }
                    catch (IOException ex)
                    {
                        _logger.Error($"IO error in reader thread: {ex.Message}");
                                                 if (!_finished)
                        {
                            _finished = true;
                            break;
                        }
                    }
                }
                else
                {
                                         Thread.Sleep(100);
                }
            }
        }
        catch (Exception ex)
        {
            _logger.Error($"Unexpected error in reader thread: {ex.Message}", ex);
        }
        finally
        {
            _logger.Info("Reader thread exiting");
        }
    }

    private void HandleResultUpdated(Response response)
    {
                 if (_client != null && response.Results != null && response.Results.Count > 0)
        {
            try
            {
                var result = response.Results[0];
                _logger.Info($"Notifying client about result update: " +
                        $"{result.Participant.Name} - {result.Points} points");

                _client.ResultsUpdated(result);
            }
            catch (TriathlonException ex)
            {
                _logger.Warn($"Error notifying client: {ex.Message}");
            }
        }
        else
        {
            _logger.Warn($"Cannot handle result update: client={_client}" +
                    $", result={(response.Results != null && response.Results.Count > 0 ? "present" : "null")}");
        }
    }

    public void Dispose()
    {
        CloseConnection();
    }
}